# Gather

In [264]:
import pandas as pd
import requests
import os
import config
from datetime import datetime

## WeRateDogs data

In [265]:
tweets = pd.read_csv('twitter-archive-enhanced.csv')

## tweet image predictions

In [266]:
if not os.path.exists('image-predictions.csv'):
    r = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv')
    with open('image-predictions.tsv', 'wb') as f:
        f.write(r.content)
        
images = pd.read_csv('image-predictions.tsv', sep='\t')

## retweet and favourites counts

In [267]:
import tweepy

auth = tweepy.OAuthHandler(config.api_key, config.api_secret)
auth.set_access_token(config.access_token, config.access_secret)

api = tweepy.API(auth)

In [268]:
if not os.path.exists('ret_fav_count.csv'):
    start_time = datetime.now()
    ret_fav_count_list = []
    deleted_tweets = []
    for tweet_id in list(tweets.tweet_id):
        try:
            tweet_info = api.get_status(tweet_id, tweet_mode='extended')
            retweet_count = tweet_info.retweet_count
            favourite_count = tweet_info.favorite_count
            ret_fav_count_list.append({'tweet_id':tweet_id,
                                       'retweet_count':retweet_count,
                                       'favourite_count':favourite_count})
        except:
            deleted_tweets.append(tweet_id)

    ret_fav_count = pd.DataFrame(ret_fav_count_list, columns=['tweet_id', 'retweet_count', 'favourite_count'])
    end_time = datetime.now()
    time_taken = (end_time - start_time).total_seconds()
    print('{} tweets are deleted.'.format(len(deleted_tweets)))
    ret_fav_count.to_csv('ret_fav_count.csv', index=False)

    print('Time taken: {} minutes'.format(round(time_taken/60, 2)))

ret_fav_count = pd.read_csv('ret_fav_count.csv')

# Assess

## tweets table

In [269]:
tweets

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

In [270]:
tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [271]:
tweets.tweet_id.duplicated().sum()

0

In [272]:
tweets[tweets['in_reply_to_status_id'].notnull()].in_reply_to_user_id

30      2.281182e+09
55      4.738443e+07
64      3.105441e+09
113     1.648776e+07
148     7.759620e+07
            ...     
2038    4.196984e+09
2149    4.196984e+09
2169    4.196984e+09
2189    2.143566e+07
2298    4.196984e+09
Name: in_reply_to_user_id, Length: 78, dtype: float64

In [273]:
tweets.source.unique()

array(['<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
       '<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>',
       '<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>',
       '<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>'],
      dtype=object)

In [274]:
len(tweets[tweets.doggo == 'None'][tweets.pupper == 'None'][tweets.puppo == 'None'][tweets.floofer == 'None'])

C:\Users\zeyad_000\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


1976

In [275]:
tweets[tweets.retweeted_status_id.notnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257                    NaN                  NaN   
32    886054160059072513                    NaN                  NaN   
36    885311592912609280                    NaN                  NaN   
68    879130579576475649                    NaN                  NaN   
73    878404777348136964                    NaN                  NaN   
...                  ...                    ...                  ...   
1023  746521445350707200                    NaN                  NaN   
1043  743835915802583040                    NaN                  NaN   
1242  711998809858043904                    NaN                  NaN   
2259  667550904950915073                    NaN                  NaN   
2260  667550882905632768                    NaN                  NaN   

                      timestamp  \
19    2017-07-21 01:02:36 +0000   
32    2017-07-15 02:45:48 +0000   
36    2017-07-13 01:35:06 +0000   
68    2017-06-26 00:13:58 +0000   
73    2017-06-24 00:09:53 +0000   
...                         ...   
1023  2016-06-25 01:52:36 +0000   
1043  2016-06-17 16:01:16 +0000   
1242  2016-03-21 19:31:59 +0000   
2259  2015-11-20 03:51:52 +0000   
2260  2015-11-20 03:51:47 +0000   

                                                 source  \
19    <a href="http://twitter.com/download/iphone" r...   
32    <a href="http://twitter.com/download/iphone" r...   
36    <a href="http://twitter.com/download/iphone" r...   
68    <a href="http://twitter.com/download/iphone" r...   
73    <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
1023  <a href="http://twitter.com/download/iphone" r...   
1043  <a href="http://twitter.com/download/iphone" r...   
1242  <a href="http://twitter.com/download/iphone" r...   
2259  <a href="http://twitter.com" rel="nofollow">Tw...   
2260  <a href="http://twitter.com" rel="nofollow">Tw...   

                                                   text  retweeted_status_id  \
19    RT @dog_rates: This is Canela. She attempted s...         8.874740e+17   
32    RT @Athletics: 12/10 #BATP https://t.co/WxwJmv...         8.860537e+17   
36    RT @dog_rates: This is Lilly. She just paralle...         8.305833e+17   
68    RT @dog_rates: This is Emmy. She was adopted t...         8.780576e+17   
73    RT @dog_rates: Meet Shadow. In an attempt to r...         8.782815e+17   
...                                                 ...                  ...   
1023  RT @dog_rates: This is Shaggy. He knows exactl...         6.678667e+17   
1043  RT @dog_rates: Extremely intelligent dog here....         6.671383e+17   
1242  RT @twitter: @dog_rates Awesome Tweet! 12/10. ...         7.119983e+17   
2259  RT @dogratingrating: Exceptional talent. Origi...         6.675487e+17   
2260  RT @dogratingrating: Unoriginal idea. Blatant ...         6.675484e+17   

      retweeted_status_user_id retweeted_status_timestamp  \
19                4.196984e+09  2017-07-19 00:47:34 +0000   
32                1.960740e+07  2017-07-15 02:44:07 +0000   
36                4.196984e+09  2017-02-12 01:04:29 +0000   
68                4.196984e+09  2017-06-23 01:10:23 +0000   
73                4.196984e+09  2017-06-23 16:00:04 +0000   
...                        ...                        ...   
1023              4.196984e+09  2015-11-21 00:46:50 +0000   
1043              4.196984e+09  2015-11-19 00:32:12 +0000   
1242              7.832140e+05  2016-03-21 19:29:52 +0000   
2259              4.296832e+09  2015-11-20 03:43:06 +0000   
2260              4.296832e+09  2015-11-20 03:41:59 +0000   

                                          expanded_urls  rating_numerator  \
19    https://twitter.com/dog_rates/status/887473957...                13   
32    https://twitter.com/dog_rates/status/886053434...                12   
36    https://twitter.com/dog_rates/status/830583320...                13   
68    https://twitt

In [276]:
tweets[tweets.in_reply_to_status_id.notnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
...                  ...                    ...                  ...   
2038  671550332464455680           6.715449e+17         4.196984e+09   
2149  669684865554620416           6.693544e+17         4.196984e+09   
2169  669353438988365824           6.678065e+17         4.196984e+09   
2189  668967877119254528           6.689207e+17         2.143566e+07   
2298  667070482143944705           6.670655e+17         4.196984e+09   

                      timestamp  \
30    2017-07-15 16:51:35 +0000   
55    2017-07-02 21:58:53 +0000   
64    2017-06-27 12:14:36 +0000   
113   2017-06-02 19:38:25 +0000   
148   2017-05-13 16:15:35 +0000   
...                         ...   
2038  2015-12-01 04:44:10 +0000   
2149  2015-11-26 01:11:28 +0000   
2169  2015-11-25 03:14:30 +0000   
2189  2015-11-24 01:42:25 +0000   
2298  2015-11-18 20:02:51 +0000   

                                                 source  \
30    <a href="http://twitter.com/download/iphone" r...   
55    <a href="http://twitter.com/download/iphone" r...   
64    <a href="http://twitter.com/download/iphone" r...   
113   <a href="http://twitter.com/download/iphone" r...   
148   <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2038  <a href="http://twitter.com/download/iphone" r...   
2149  <a href="http://twitter.com/download/iphone" r...   
2169  <a href="http://twitter.com/download/iphone" r...   
2189  <a href="http://twitter.com/download/iphone" r...   
2298  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
30    @NonWhiteHat @MayhewMayhem omg hello tanner yo...                  NaN   
55    @roushfenway These are good dogs but 17/10 is ...                  NaN   
64                      @RealKentMurphy 14/10 confirmed                  NaN   
113   @ComplicitOwl @ShopWeRateDogs &gt;10/10 is res...                  NaN   
148   @Jack_Septic_Eye I'd need a few more pics to p...                  NaN   
...                                                 ...                  ...   
2038  After 22 minutes of careful deliberation this ...                  NaN   
2149  After countless hours of research and hundreds...                  NaN   
2169  This is Tessa. She is also very pleased after ...                  NaN   
2189                     12/10 good shit Bubka\n@wane15                  NaN   
2298  After much debate this dog is being upgraded t...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
30                         NaN                        NaN   
55                         NaN                        NaN   
64                         NaN                        NaN   
113                        NaN                        NaN   
148                        NaN                        NaN   
...                        ...                        ...   
2038                       NaN                        NaN   
2149                       NaN                        NaN   
2169                       NaN                        NaN   
2189                       NaN                        NaN   
2298                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
30                                                  NaN                12   
55                                                  NaN                17   
64                                                  NaN                14   
113                

In [277]:
tweets[tweets.in_reply_to_status_id.notnull()].text[55]

'@roushfenway These are good dogs but 17/10 is an emotional impulse rating. More like 13/10s'

In [278]:
tweets[tweets.in_reply_to_status_id.notnull()].in_reply_to_status_id[30]

8.862663570751283e+17

In [279]:
tweets.rating_numerator.sort_values(ascending=False)

979     1776
313      960
189      666
188      420
2074     420
        ... 
2338       1
2261       1
2335       1
1016       0
315        0
Name: rating_numerator, Length: 2356, dtype: int64

In [280]:
tweets[tweets.rating_numerator > 15]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
55    881633300179243008           8.816070e+17         4.738443e+07   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
290   838150277551247360           8.381455e+17         2.195506e+07   
313   835246439529840640           8.352460e+17         2.625958e+07   
340   832215909146226688                    NaN                  NaN   
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
695   786709082849828864                    NaN                  NaN   
763   778027034220126208                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
979   749981277374128128                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1712  680494726643068929                    NaN                  NaN   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2074  670842764863651840                    NaN                  NaN   

                      timestamp  \
55    2017-07-02 21:58:53 +0000   
188   2017-04-22 19:15:32 +0000   
189   2017-04-22 19:05:32 +0000   
290   2017-03-04 22:12:52 +0000   
313   2017-02-24 21:54:03 +0000   
340   2017-02-16 13:11:49 +0000   
433   2017-01-15 17:52:40 +0000   
516   2016-12-19 23:06:23 +0000   
695   2016-10-13 23:23:56 +0000   
763   2016-09-20 00:24:34 +0000   
902   2016-07-28 01:00:57 +0000   
979   2016-07-04 15:00:45 +0000   
1120  2016-05-13 16:15:54 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1663  2016-01-01 06:22:03 +0000   
1712  2015-12-25 21:06:00 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   
2074  2015-11-29 05:52:33 +0000   

                                                 source  \
55    <a href="http://twitter.com/download/iphone" r...   
188   <a href="http://twitter.com/download/iphone" r...   
189   <a href="http://twitter.com/download/iphone" r...   
290   <a href="http://twitter.com/download/iphone" r...   
313   <a href="http://twitter.com/download/iphone" r...   
340   <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
695   <a href="http://twitter.com/download/iphone" r...   
763   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
979   <a href="https://about.twitter.com/products/tw...   
1120  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1433

In [407]:
tweets.name.value_counts()

None        745
a           55 
Charlie     12 
Oliver      11 
Cooper      11 
            .. 
Kayla       1  
Willy       1  
Vinscent    1  
Fillup      1  
Harlso      1  
Name: name, Length: 957, dtype: int64

In [409]:
tweets[tweets.name=='a'].text

56      Here is a pupper approaching maximum borkdrive. Zooming at never before seen speeds. 14/10 paw-inspiring af \n(IG: puffie_the_chow) https://t.co/ghXBIIeQZF
649     Here is a perfect example of someone who has their priorities in order. 13/10 for both owner and Forrest https://t.co/LRyMrU7Wfq                           
801     Guys this is getting so out of hand. We only rate dogs. This is a Galapagos Speed Panda. Pls only send dogs... 10/10 https://t.co/8lpAGaZRFn               
1002    This is a mighty rare blue-tailed hammer sherk. Human almost lost a limb trying to take these. Be careful guys. 8/10 https://t.co/TGenMeXreW               
1004    Viewer discretion is advised. This is a terrible attack in progress. Not even in water (tragic af). 4/10 bad sherk https://t.co/L3U0j14N5R                 
1017    This is a carrot. We only rate dogs. Please only send in dogs. You all really should know this by now ...11/10 https://t.co/9e48aPrBm2                     
1049    This is 

## images table

In [281]:
images

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

In [282]:
images.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [283]:
images[images.p1_dog == True]

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2068  890971913173991426  https://pbs.twimg.com/media/DF1eOmZXUAALUcq.jpg   
2069  891087950875897856  https://pbs.twimg.com/media/DF3HwyEWsAABqE6.jpg   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   

      img_num                        p1   p1_conf  p1_dog                  p2  \
0           1    Welsh_springer_spaniel  0.465074    True              collie   
1           1                   redbone  0.506826    True  miniature_pinscher   
2           1           German_shepherd  0.596461    True            malinois   
3           1       Rhodesian_ridgeback  0.408143    True             redbone   
4           1        miniature_pinscher  0.560311    True          Rottweiler   
...       ...                       ...       ...     ...                 ...   
2068        1               Appenzeller  0.341703    True       Border_collie   
2069        1  Chesapeake_Bay_retriever  0.425595    True       Irish_terrier   
2070        2                    basset  0.555712    True    English_springer   
2072        1                 Chihuahua  0.716012    True            malamute   
2073        1                 Chihuahua  0.323581    True            Pekinese   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2068  0.199287    True                    ice_lolly  0.193548   False  
2069  0.116317    True              Indian_elephant  0.076902   False  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  

[1532 rows x 12 columns]

## retweet and favourite counts

In [284]:
ret_fav_count

tweet_id  retweet_count  favourite_count
0     892420643555336193           7435            35260
1     892177421306343426           5527            30516
2     891815181378084864           3648            22945
3     891689557279858688           7611            38551
4     891327558926688256           8194            36828
...                  ...            ...              ...
1544  680805554198020098            640             2076
1545  680801747103793152            786             2296
1546  680798457301471234            992             2751
1547  680609293079592961            668             2535
1548  680583894916304897           1274             3502

[1549 rows x 3 columns]

In [285]:
ret_fav_count.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1549 entries, 0 to 1548
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   tweet_id         1549 non-null   int64
 1   retweet_count    1549 non-null   int64
 2   favourite_count  1549 non-null   int64
dtypes: int64(3)
memory usage: 36.4 KB


In [286]:
(ret_fav_count.retweet_count > ret_fav_count.favourite_count).sum()

139

In [287]:
ret_fav_count[ret_fav_count.retweet_count > ret_fav_count.favourite_count].sample(10)

tweet_id  retweet_count  favourite_count
582   798673117451325440           5503                0
559   800855607700029440           1593                0
313   833732339549220864            213                0
636   791821351946420224          20866                0
556   801127390143516673           3914                0
269   839290600511926273            139                0
460   816014286006976512           5167                0
580   798686750113755136           2309                0
1080  711998809858043904            123                0
430   819015337530290176          36318                0

## Quality

tweets table
- in_reply_to_status_user_id and in_reply_to_user_id columns are float, and have e+17 and a dot.
- timestamp column is object
- retweets act as duplicates
- numerator of index 55 is not 17, but 13 instead
- denominator does not have 10 as the only value, rating score should be one column
- rating score has outliers (some of them because of decimals in the numerators)
- expanded_urls column has repeated values if the tweet has more than one photo
- name column has 'a' and 'the' values that should be replaced with 'None'

images table
- 3 falses in the p#_dog value
- 3 breeds for the same picture

ret_fav_count table
- 139 tweets have 0 value in the favourite count, that's because they are related to retweets

## Tidiness

tweets table
- doggo, floofer, pupper, and puppo are 4 attributes that should be combined to be just one attribute name 'stage'

images table
- one breed column should be created
- breed column should belong to tweets table

ret_fav_count table
- it should belong to the tweets table

# Clean

## Quality

### tweets table

In [288]:
tweets_clean = tweets.copy()

#### Define

- change in_reply_to_status_user_id and in_reply_to_user_id columns from float to string
- remove "e+17" and the dot
- replace NaN values with "N/A"

#### Code

In [289]:
tweets_clean['in_reply_to_status_id'] = tweets_clean['in_reply_to_status_id'].astype(str)
tweets_clean['in_reply_to_status_id'] = tweets_clean['in_reply_to_status_id'].apply(lambda x: x.replace('e+17', '')\
                                                                                               .replace('.', '')\
                                                                                               .replace('nan', 'N/A'))

In [290]:
tweets_clean['in_reply_to_user_id'] = tweets_clean['in_reply_to_user_id'].astype(str)
tweets_clean['in_reply_to_user_id'] = tweets_clean['in_reply_to_user_id'].apply(lambda x: x.replace('e+17', '')\
                                                                                               .replace('.', '')\
                                                                                               .replace('nan', 'N/A'))

#### Test

In [291]:
tweets_clean[tweets_clean.in_reply_to_status_id != 'N/A'].in_reply_to_status_id

30      8862663570751283
55      8816070373140521
64      8795538273341727
113     8707262027424932
148     8634256455687741
              ...       
2038    6715448741650022
2149    6693543826270495
2169    6678064545737605
2189    6689207171325829
2298    6670655355705508
Name: in_reply_to_status_id, Length: 78, dtype: object

In [292]:
tweets_clean[tweets_clean.in_reply_to_user_id != 'N/A'].in_reply_to_user_id

30      22811816000
55        473844300
64      31054407460
113       164877600
148       775962000
           ...     
2038    41969838350
2149    41969838350
2169    41969838350
2189      214356580
2298    41969838350
Name: in_reply_to_user_id, Length: 78, dtype: object

#### Define

- change timestamp column from object to datetime

#### Code

In [293]:
tweets_clean['timestamp'] = pd.to_datetime(tweets_clean['timestamp'])

#### Test

In [294]:
tweets_clean['timestamp'].dtype.name

'datetime64[ns, UTC]'

#### Define

- remove retweets as they act as duplicates
- drop retweeted_status_user_id and retweeted_status_id columns

#### Code

In [364]:
idx_to_drop = tweets_clean[tweets_clean['retweeted_status_id'].notnull()].index
tweets_clean.drop(index=idx_to_drop, inplace=True)

tweets_clean.drop(columns=['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'],
                  inplace=True)

#### Test

In [296]:
tweets_clean[tweets_clean.index.isin(idx_to_drop)]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

#### Define

- change numerator of index 55 is not 17, but 13 instead

#### Code

In [297]:
tweets_clean.loc[55].rating_numerator

17

In [298]:
tweets_clean.loc[55, 'rating_numerator'] = 13

#### Test

In [299]:
tweets_clean.loc[55].rating_numerator

13

#### Define

- create a rating_score column by dividing rating_numerator by rating_denominator
- drop rating_numerator and rating_denominator columns

#### Code

In [300]:
tweets_clean['rating_score'] = tweets_clean['rating_numerator'] / tweets['rating_denominator']
tweets_clean.drop(columns=['rating_numerator', 'rating_denominator'], inplace=True)

#### Test

In [301]:
tweets_clean.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'name', 'doggo',
       'floofer', 'pupper', 'puppo', 'rating_score'],
      dtype='object')

In [302]:
tweets_clean.rating_score.head()

0    1.3
1    1.3
2    1.2
3    1.3
4    1.2
Name: rating_score, dtype: float64

#### Define

- remove outliers in rating_score column (some of them had decimals in the numerators)
    1. since WeRateDogs rate dogs mostly 10/10, then values less than 1 will be substituted as 1
    2. remove values more than 1.5

#### Code

In [303]:
tweets_clean['rating_score'] = tweets_clean['rating_score'].apply(lambda x: 1 if x < 1 else x)

In [304]:
idx_to_drop = tweets_clean[tweets_clean['rating_score'] > 1.5].index
tweets_clean.drop(index=idx_to_drop, inplace=True)

#### Test

In [305]:
tweets_clean[tweets_clean['rating_score'] < 1]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, name, doggo, floofer, pupper, puppo, rating_score]
Index: []

In [306]:
tweets_clean[tweets_clean['rating_score'] > 1.5]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, name, doggo, floofer, pupper, puppo, rating_score]
Index: []

#### Define

- make the expanded_urls with only one value per tweet

#### Code

In [307]:
tweets_clean['expanded_urls'] = tweets_clean['expanded_urls'].dropna().apply(lambda x: x[0] if len(x.split(',')) > 1 else x)

#### Test

In [308]:
tweets_clean.expanded_urls.values

array(['https://twitter.com/dog_rates/status/892420643555336193/photo/1',
       'https://twitter.com/dog_rates/status/892177421306343426/photo/1',
       'https://twitter.com/dog_rates/status/891815181378084864/photo/1',
       ...,
       'https://twitter.com/dog_rates/status/666033412701032449/photo/1',
       'https://twitter.com/dog_rates/status/666029285002620928/photo/1',
       'https://twitter.com/dog_rates/status/666020888022790149/photo/1'],
      dtype=object)

#### Define

- replace 'a' or 'the' with 'None' in name column

#### Code

In [429]:
tweets_clean['name'] = tweets_clean['name'].apply(lambda x: 'None' if ((x == 'a') or (x == 'the')) else x)

#### Test

In [411]:
tweets[tweets_clean['name']=='a']

C:\Users\zeyad_000\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [430]:
tweets[tweets_clean['name']=='the']

C:\Users\zeyad_000\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

### images table

In [309]:
images_clean = images.copy()

#### Define

- remove rows with 3 falses in the p#_dog value

#### Code

In [310]:
idx_to_drop = images_clean[images_clean.p1_dog == False][images_clean.p2_dog == False][images_clean.p3_dog == False].index
images_clean.drop(index=idx_to_drop, inplace=True)

C:\Users\zeyad_000\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


#### Test

In [311]:
images_clean[images_clean.p1_dog == False][images_clean.p2_dog == False][images_clean.p3_dog == False]

C:\Users\zeyad_000\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog]
Index: []

### retweet and favourite counts table

#### Define

- remove records with favourite_count equal to zero

#### Code

In [312]:
idx_to_drop = ret_fav_count[ret_fav_count['favourite_count']==0].index
ret_fav_count.drop(index=idx_to_drop, inplace=True)

#### Test

In [313]:
ret_fav_count[ret_fav_count['favourite_count']==0]

Empty DataFrame
Columns: [tweet_id, retweet_count, favourite_count]
Index: []

## Tidiness

### tweets table

#### Define

create 'stage' column for all 4 columns; doggo, floofer, pupper, and puppo
1. specify columns that will not change
2. use melt function with variable name 'dog' and value 'stage'
3. drop dog column
4. drop duplicates

#### Code

In [315]:
cols = list(tweets_clean.columns)
cols_remove = ['doggo', 'floofer', 'pupper', 'puppo']
for col in cols_remove:
    cols.remove(col)
    
tweets_clean = pd.melt(tweets_clean, id_vars=cols, var_name='dog', value_name='stage')
tweets_clean.drop(columns='dog', inplace=True)
tweets_clean.drop_duplicates(inplace=True)

#### Test

In [326]:
tweets_clean.stage.sample(20)

151       None
1776      None
813       None
995       None
1928      None
1963      None
409      doggo
5882    pupper
437      doggo
1474      None
761       None
414       None
921       None
1853      None
6958     puppo
1835      None
617       None
1014      None
1998      None
6137    pupper
Name: stage, dtype: object

### images table

#### Define

- select breed value depending on boolean values in p#_dog

#### Code

In [344]:
breed_p1 = images_clean[images_clean['p1_dog']==True][['tweet_id', 'p1']]
breed_p1.rename(columns={'p1':'breed'}, inplace=True)

breed_p2 = images_clean[images_clean['p1_dog']==False][images_clean['p2_dog']==True][['tweet_id', 'p2']]
breed_p2.rename(columns={'p2':'breed'}, inplace=True)

breed_p3 = images_clean[images_clean['p1_dog']==False][images_clean['p2_dog']==False][images_clean['p3_dog']==True][['tweet_id', 'p3']]
breed_p3.rename(columns={'p3':'breed'}, inplace=True)

breed = pd.concat([breed_p1, breed_p2, breed_p3])

C:\Users\zeyad_000\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  after removing the cwd from sys.path.
C:\Users\zeyad_000\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  import sys


#### Test

In [347]:
breed.head()

tweet_id                   breed
0  666020888022790149  Welsh_springer_spaniel
1  666029285002620928                 redbone
2  666033412701032449         German_shepherd
3  666044226329800704     Rhodesian_ridgeback
4  666049248165822465      miniature_pinscher

#### Define

- combine breeds column to tweets table

#### Code

In [350]:
tweets_clean = tweets_clean.merge(breed, on='tweet_id', how='left')

#### Test

In [353]:
tweets_clean.sample(2)

tweet_id in_reply_to_status_id in_reply_to_user_id  \
2272  845306882940190720                   N/A                 N/A   
1721  674422304705744896                   N/A                 N/A   

                     timestamp  \
2272 2017-03-24 16:10:40+00:00   
1721 2015-12-09 02:56:22+00:00   

                                                 source  \
2272  <a href="http://twitter.com/download/iphone" r...   
1721  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2272  This is Pickles. She's a silly pupper. Thinks ...                  NaN   
1721  This is Ava. She doesn't understand flowers. 1...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2272                       NaN                        NaN   
1721                       NaN                        NaN   

                                          expanded_urls     name  \
2272  https://twitter.com/dog_rates/status/845306882...  Pickles   
1721  https://twitter.com/dog_rates/status/674422304...      Ava   

      rating_score   stage                breed  
2272           1.2  pupper  Irish_water_spaniel  
1721           1.2    None     golden_retriever

#### Define

- combine ret_fav_counts table to tweets table

#### Code

In [354]:
tweets_clean = tweets_clean.merge(ret_fav_count, on='tweet_id', how='left')

#### Test

In [355]:
tweets_clean.head(2)

tweet_id in_reply_to_status_id in_reply_to_user_id  \
0  892420643555336193                   N/A                 N/A   
1  892177421306343426                   N/A                 N/A   

                  timestamp  \
0 2017-08-01 16:23:56+00:00   
1 2017-08-01 00:17:27+00:00   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls     name  rating_score  \
0  https://twitter.com/dog_rates/status/892420643...  Phineas           1.3   
1  https://twitter.com/dog_rates/status/892177421...    Tilly           1.3   

  stage      breed  retweet_count  favourite_count  
0  None        NaN         7435.0          35260.0  
1  None  Chihuahua         5527.0          30516.0

## Final Adjustments tweets table

#### Define

- convert retweet_count and favourite_count columns from float to int

#### Code

In [362]:
tweets_clean['retweet_count'] = tweets_clean['retweet_count'].astype('Int64')
tweets_clean['favourite_count'] = tweets_clean['favourite_count'].astype('Int64')

#### Test

In [363]:
tweets_clean.dtypes

tweet_id                                    int64
in_reply_to_status_id                      object
in_reply_to_user_id                        object
timestamp                     datetime64[ns, UTC]
source                                     object
text                                       object
retweeted_status_id                       float64
retweeted_status_user_id                  float64
retweeted_status_timestamp                 object
expanded_urls                              object
name                                       object
rating_score                              float64
stage                                      object
breed                                      object
retweet_count                               Int64
favourite_count                             Int64
dtype: object

# Store

In [431]:
tweets_clean.to_csv('twitter_arhive_master.csv', index=False)

# Analysis

In [432]:
df = pd.read_csv('twitter_arhive_master.csv')
df.dtypes

tweet_id                 int64  
in_reply_to_status_id    float64
in_reply_to_user_id      float64
timestamp                object 
source                   object 
text                     object 
expanded_urls            object 
name                     object 
rating_score             float64
stage                    object 
breed                    object 
retweet_count            float64
favourite_count          float64
dtype: object

When data is being stored and read again, columns types change. Add a function that takes care of the types of the columns.

In [433]:
def change_types(df):
    df.timestamp = pd.to_datetime(df.timestamp)
    df.retweet_count = df.retweet_count.astype('Int64')
    df.favourite_count = df.favourite_count.astype('Int64')
    return df

df = change_types(df)

## Popular Breeds

In [438]:
list(df.breed.value_counts().head().index)

['golden_retriever', 'Labrador_retriever', 'Pembroke', 'Chihuahua', 'pug']

## Duration

In [393]:
duration = df.timestamp.max() - df.timestamp.min()
print('Duration: {} years'.format(round(duration.days/365, 1)))

Duration: 1.7 years


## Number of Tweets

In [398]:
print('Number of Tweets: {}'.format(df.tweet_id.nunique()))

Number of Tweets: 2165


## Popular Names

In [436]:
list(df[df.name!='None'].name.value_counts().head().index)

['Cooper', 'Charlie', 'Lucy', 'Oliver', 'Penny']

## Popular Stage

In [442]:
df[df.stage!='None'].stage.value_counts()

pupper     233
doggo      87 
puppo      25 
floofer    10 
Name: stage, dtype: int64